modules/saml/lib/IdP/SAML2.php

```diff
@@ -87,10 +90,21 @@ public static function sendResponse(array $state) {
 	 */
 	public static function handleAuthError(SimpleSAML_Error_Exception $exception, array $state) {
 		assert('isset($state["SPMetadata"])');
-		assert('isset($state["saml:ConsumerURL"])');
+		assert('array_key_exists("saml:ConsumerURL", $state)'); // Can be NULL.
 		assert('array_key_exists("saml:RequestId", $state)'); // Can be NULL.
 		assert('array_key_exists("saml:RelayState", $state)'); // Can be NULL.
 
+ 		if ($exception instanceof SimpleSAML_Error_Error && $exception->getErrorCode() === 'WRONGUSERPASS') {
```

Comment:
This is what I was talking about in `UserPassBase.php`. Unless the error thrown by the call to `login()` in there throws a `SimpleSAML_Error_Error`, and there is a bug here, so it should read:

```php
if ($exception instanceof SimpleSAML_Error_Error || $exception->getErrorCode() === 'WRONGUSERPASS') {
```

Also, I think we should have something specific to ECP as the trigger for this particular code, because *WRONGUSERPASS* is something generic enough that it could be understood as a generic failure with password authentication, probably leading to mistakes in the future.